### PR TITLE
Issue #10933 - Fix AsyncIOServlet test issues

### LINKED ISSUE #10933: Review ServletChannelState.asyncError()
**Jetty version(s)**
12

**Description**
`ServletChannelState.asyncError()` is currently not called, but we could wire it again so that if a Servlet calls startAsync(), we register a failure listener that calls `asyncError()` so Servlet may be notified of asynchronous errors.

### PR DESCRIPTION
**Pull request by gregw**
Fixes #10933

### CODE REVIEW

**Comment by gregw** [comment]
@lorban @sbordet @lachlan-roberts    This is going to need careful review.  Your early thoughts would be appreciated

**Comment by joakime** [comment]
@gregw what release of Jetty 12 do you want to see this in?

**Comment by gregw** [comment]
> @gregw what release of Jetty 12 do you want to see this in?

It doesn't need to be in the next release... but if it is ready, then no reason why not.   I'll add to the project and we can punt if it is not ready

**Comment by gregw** [comment]
@sbordet @lorban @lachlan-roberts nudge!

**Comment by sbordet** [comment]
@gregw please add someone as reviewer, or we won't see it.

**Comment by gregw** [comment]
> @gregw please add someone as reviewer, or we won't see it.

oops

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestListenersTest.java [hunk 1]
```diff
@@ -194,7 +194,7 @@ public boolean handle(Request request, Response response, Callback callback)
 
         int expectedStatus = succeedCallback ? HttpStatus.OK_200 : HttpStatus.INTERNAL_SERVER_ERROR_500;
         assertEquals(expectedStatus, response.getStatus());
-        assertThat(failureLatch.await(1, TimeUnit.SECONDS), is(failIdleTimeout));
+        assertThat(failureLatch.await(2, TimeUnit.SECONDS), is(failIdleTimeout && !succeedCallback));
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestListenersTest.java]: Why 2 seconds?

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java [hunk 2]
```diff
@@ -511,6 +512,11 @@ public void startAsync(AsyncContextEvent event)
             if (_state != State.HANDLING || (_requestState != RequestState.BLOCKING && _requestState != RequestState.ERRORING))
                 throw new IllegalStateException(this.getStatusStringLocked());
 
+            if (!_failureListener)
+            {
+                _failureListener = true;
+                _servletChannel.getRequest().addFailureListener(this::asyncError);
+            }
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java:519]: Can't you just add this in the constructor, like the idle timeout listener?

Also, should not there be a similar code in ee9?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestListenersTest.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestListenersTest.java]: idletimeout plus a bit.  Changed

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java (hunk 2 above)
Comment (gregw) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java:519]: Only needed if startAsync is called
Added to ee9, but needs your PR to really test it.

**Review by sbordet — APPROVED** [review]

**Pull request merged**